widgetschema.json

```diff
@@ -0,0 +1,244 @@
+        },
+        "src": {
+            "type": "string",
+            "description": "URL of the widget source code"
+        },
+        "needsDataSource": {
+            "type": "boolean",
+            "description": "Indicates if the widget needs a data source"
+        },
+        "needsInstance": {
+            "oneOf": [{ "type": "boolean" }, { "$ref": "#/definitions/dataType" }]
+        },
+        "typesSupported": {
+            "type": "array",
+            "items": { "$ref": "#/definitions/dataType" }
+        },
+        "needsDocument": {
+            "type": "string"
+        },
+        "icon": {
+            "type": "object",
+            "required": ["src"],
+            "additionalProperties": false,
+            "properties": {
+                "src": {
+                    "type": "string"
+                }
+            }
+        },
+        "colspan": {
+            "type": "integer",
+            "minimum": 1,
+            "maximum": 12
+        },
+        "allowSticky": {
+            "type": "boolean"
+        },
+        "deprecated": {
+            "type": "boolean"
+        },
+        "properties": {
+            "type": "array",
+            "uniqueItems": true,
+            "description": "Array of properties that the widget supports",
+            "items": { "$ref": "#/definitions/widgetProperty" }
+        },
+        "$schema": { "type": "string" }
+    },
+
+    "definitions": {
+        "dataType": {
+            "enum": ["documents", "objectInstances", "documentVersions", "relatedObjectInstances"]
+        },
+        "widgetSettingOptionalValue": {
+            "type": "object",
+            "required": ["value"],
+            "additionalProperties": false,
+            "properties": {
+                "value": { "type": "boolean" },
+                "initialValue": {}
+            }
+        },
+        "choices": {
+            "type": "array",
+            "uniqueItems": true,
+            "description": "Array of choices for the property",
+            "items": {
+                "$ref": "#/definitions/choice"
+            }
+        },
+        "choice": {
+            "type": "object",
+            "required": ["label", "value"],
+            "additionalProperties": false,
+            "properties": {
+                "value": {
+                    "type": "string",
+                    "description": "Value of the choice"
+                },
+                "label": {
+                    "type": "string",
+                    "description": "Label of the choice"
+                }
+            }
+        },
+        "api": {
+            "type": "object",
+            "required": ["url", "resultMapping"],
+            "additionalProperties": false,
+            "properties": {
+                "url": {
+                    "type": "string",
+                    "description": "URL of the API"
+                },
+                "resultMapping": {
```

Comment:
For my clarification, what would be an example `resultMapping`?  I'm trying to figure out how `label` and `value` are used, and what the purpose of `additionalProperties` is.

Reply:
Value is the value that gets saved and is passed in as props. Label is what is what gets displayed in the dropdown. In all the uses that I have seen label is always set to `name` and value is always set to `id`

`additionalProperties: true` would allow properties to be added with any name even if not added to the schema. For example if `additionalProperties` was set to true the below would be valid even though food is not part of the schema properties.

```
 "api": {
        "url": "/objects/$_objectId/properties?filter={\"where\":{\"or\":[{\"type\":\"string\"},{\"type\":\"richText\"}]}}",
        "resultMapping": {
          "value": "id",
          "label": "name"
        },
      "food": "pizza"
      }
```

Reply:
I don't mean what `additionalProperties` does in general.  I mean why are additional properties allowed?

As far as `label` and `value` go, it sounds like these are used to specify properties of the result which are then used as the option labels and values respectively.  Even if structurally it's the same as `Choice`, I think this should be its own type because it's not really the same thing as `Choice`, e.g. the descriptions for `label` and `value` would be different since these are specifying the label/value mapping configuration.

Reply:
My bad. It doesn't seem like `"additionalProperties": {"type": "string"}` is doing anything at all. I will remove it